File: day8/part2.py
import numpy as np
import time

start = time.time()

class Node:
    def __init__(self, value: str):
        self.value = value
        
    def setL(self, left_node):
        self.l = left_node
    
    def setR(self, right_node):
        self.r = right_node
    
    def __repr__(self):
        return self.value 
    # + " " + self.l.value + " " + self.r.value
        
with open("data.txt") as data:
    lines = data.readlines()
    instructions = lines[0].strip()
    
    node_data = {}
    nodes = {}
    curr_nodes = []
    
    for line in lines[2:]:
        node = line[:3]
        left = line[7:10]
        right = line[12:15]
        
        node_data[node] = {"L": left, "R": right}
        nodes[node] = Node(node)
        if node[2] == "A":
            curr_nodes.append(nodes[node])
        
    for node_val, node in nodes.items():
        node.setL(nodes[node_data[node_val]["L"]])
        node.setR(nodes[node_data[node_val]["R"]])
        
    count = 0
    def all_end_in_Z():
        for node in curr_nodes: 
            if node.value[2] != "Z":
                return False
        return True
        
    counts = [0]*len(curr_nodes)
    counts_set = [0]*len(curr_nodes)
    while not sum(counts_set) == len(curr_nodes):        
        instruction = instructions[count % len(instructions)]
        for i in range(len(curr_nodes)):
            if curr_nodes[i].value[2] == "Z" and counts_set[i] == 0:
                counts[i] = count
                counts_set[i] = 1
            elif counts[i] == 0:
                if instruction == "L":
                    curr_nodes[i] = curr_nodes[i].l
                else:
                    curr_nodes[i] = curr_nodes[i].r
        count+=1
    print(counts)
    # the only reason you can use lcm here is due to luck.
    # there is really nothing in the problem indicating that each path from starting A to first Z
    # is the exit node it will eventually hit to exit, which LCM assumes. In fact, each exit isn't even a loop back to the starting node.
    print(np.lcm.reduce(counts))

print(time.time() - start)

File: day8/Part2.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.Map.Entry;

class Part2 {

    // class Node {
    //     String value;
    //     Node left;
    //     Node right;

    //     Node(String value){
    //         this.value = value;
    //     }

    //     void setLeftNode(Node left) {
    //         this.left = left;
    //     }

    //     void setRightNode(Node right) {
    //         this.right = right;
    //     }
    // /**
    //  * @param args
    //  */
    // }

    public static void main(String[] args) {
        try {
            File fileObj = new File("data.txt");
            Scanner scanner = new Scanner(fileObj);
            HashMap<String, String[]> graph = new HashMap<>();

            String instructions = scanner.nextLine();
            scanner.nextLine();
            while(scanner.hasNextLine()) {
                String line = scanner.nextLine();
                String node_val = line.substring(0,3);
                String left = line.substring(7,10);
                String right = line.substring(12,15);
                String[] lr = {left, right};
                graph.put(node_val, lr);
            }
            
            ArrayList<String> startNodes = new ArrayList<>();

            for(Entry<String, String[]> entry : graph.entrySet()) {
                    String key = entry.getKey();
                    if (key.charAt(2) == 'A'){
                        startNodes.add(key);
                    }
            }
            int count = getFinishCount(graph, instructions, startNodes);
            System.out.println(count);
        } catch(FileNotFoundException e){
        } 
    }

    static boolean allEndInZ(List<String> nodes) {
        for (String node: nodes) {
            if (!(node.charAt(2) == 'Z')) {
                return false;
            }
        }
        return true;
    }

    private static int getFinishCount(HashMap<String, String[]> graph, String instructions, ArrayList<String> currNodes) {
        int count = 0;
        Character instruction;
        int instruction_len = instructions.length();
        // System.out.println(currNodes);
        while(!allEndInZ(currNodes)) {
            instruction = instructions.charAt(count % instruction_len);
            // System.out.println(instruction);
            for (int i = 0; i < currNodes.size(); i++) {
                if (instruction == 'L') {
                    currNodes.set(i, graph.get(currNodes.get(i))[0]);
                } else {
                    currNodes.set(i, graph.get(currNodes.get(i))[1]);
                }
            }
            // System.out.println(currNodes);
            count++;
        }
        return count;
    }
}